internal/xgettext/xgettext.go

```diff
@@ -0,0 +1,469 @@
+		name = e.Name
+	case *ast.SelectorExpr:
+		name = e.Sel.Name
+		if ident, ok := e.X.(*ast.Ident); ok {
+			pkg = ident.Name
+		}
+	default:
+		return false
+	}
+
+	if name != k.name {
+		return false
+	}
+	// If the keyword includes a package qualifier, make sure it matches
+	return k.pkg == "" || k.pkg == pkg
+}
+
+func (k *Keyword) Extract(call *ast.CallExpr) (msg Message, err error) {
+	if k.msgid >= len(call.Args) {
+		return Message{}, ErrOutOfRange
+	}
+	msg.msgid, err = stringConstant(call.Args[k.msgid])
+	if err != nil {
+		return Message{}, err
+	}
+	if k.msgidPlural >= 0 {
+		if k.msgidPlural >= len(call.Args) {
+			return Message{}, ErrOutOfRange
+		}
+		msg.msgidPlural, err = stringConstant(call.Args[k.msgidPlural])
+		if err != nil {
+			return Message{}, err
+		}
+	}
+	if k.msgContext >= 0 {
+		if k.msgContext >= len(call.Args) {
+			return Message{}, ErrOutOfRange
+		}
+		msg.msgContext, err = stringConstant(call.Args[k.msgContext])
+		if err != nil {
+			return Message{}, err
+		}
+	}
+	return msg, nil
+}
+
+type Message struct {
+	msgid       string
+	msgidPlural string
+	msgContext  string
+}
+
+func (m *Message) Less(other *Message) bool {
+	if m.msgid != other.msgid {
+		return m.msgid < other.msgid
+	}
+	if m.msgidPlural != other.msgidPlural {
+		return m.msgidPlural < other.msgidPlural
+	}
+	return m.msgContext < other.msgContext
+}
+
+type Location struct {
+	file     string
+	line     int
+	comments string
+	format   bool
+}
+
+type visitor struct {
+	*Extractor
+
+	fset *token.FileSet
+	file *ast.File
+}
+
+func commentGroupContent(cg *ast.CommentGroup) string {
+	var lines []string
+	for _, comment := range cg.List {
+		for _, line := range strings.Split(comment.Text, "\n") {
+			line = strings.TrimPrefix(line, "//")
+			line = strings.TrimPrefix(line, "/*")
+			line = strings.TrimSuffix(line, "*/")
+			line = strings.TrimSpace(line)
+			if line != "" {
+				lines = append(lines, "#. "+line+"\n")
+			}
+		}
+	}
+	return strings.Join(lines, "")
+}
+
+func (v *visitor) findCommentsBefore(pos token.Position) string {
+	for i := len(v.file.Comments) - 1; i >= 0; i-- {
+		cg := v.file.Comments[i]
+		cgPos := v.fset.Position(cg.End())
+		if cgPos.Line+1 == pos.Line {
+			return commentGroupContent(cg)
+		}
+	}
+	return ""
+}
+
+func (v *visitor) Visit(node ast.Node) ast.Visitor {
+	// We're only interested in calls
+	call, ok := node.(*ast.CallExpr)
+	if !ok {
+		return v
+	}
+
+	for _, k := range v.Keywords {
+		if !k.Match(call) {
+			continue
+		}
+
+		msg, err := k.Extract(call)
+		if err != nil {
+			break
+		}
+
+		pos := v.fset.Position(node.Pos())
+		var comments string
+		if len(v.CommentTags) != 0 {
+			comments = v.findCommentsBefore(pos)
+			keep := false
+			for _, tag := range v.CommentTags {
+				if strings.HasPrefix(comments, "#. "+tag) {
+					keep = true
+					break
+				}
+			}
+			if !keep {
+				comments = ""
+			}
+		}
+
+		v.Messages[msg] = append(v.Messages[msg], Location{
+			file:     pos.Filename,
```

Comment:
Should the location be relative to the current directory or the root of the repository the file is in? That's what GNU gettext does.

Writing full paths will keep introducing changes when multiple people/systems are involved in generating the `.pot` file.